ATM 3DVAR with and without IAU

NeilBarton-NOAA opened the pull request:
This is the first of a series of PRs to addresses issue https://github.com/NOAA-EMC/global-workflow/issues/947 (cycling the S2SW app with 3DVAR). Using the ATM app, 3DVAR with IAU did not cycle out-of-the-box. This PR includes changes to successfully cycle ATM with 3DVAR with and without IAU.

Refs: #947 

Changes include:
- config.base.emc.dyn for 3DVAR IAU options
- parsing_namelists_FV3.sh to filter IAU increments when using 3DVAR

**Type of change**

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

**How Has This Been Tested?**
<!-- Please describe the tests that you ran to verify your changes and on the platforms these tests were conducted. -->
<!-- Provide instructions so we can reproduce. -->
<!-- Please also list any relevant details for your test configuration -->

Use app ` ATM` and `DOHYBVAR="NO"` with `DOIAU="NO"` and `DOIAU="YES"`. Resolution tested was `384`. 

I cycled with a start date of `2020040100`  until `2020040400`. The restarts are located on hera at
`/scratch2/NCEPDEV/stmp1/Neil.Barton/ICs/2020040100/C384/control`

I also rested the default config with  `DOHYBVAR="YES"` .

At one point I ran into an issue with fit2obs and gdaspostanl. However, the latest pull from develop fixed these issues for me. 

- [x] Cycled test on Hera
- [x] Cycled test on Orion (tested before the latest pull from develop) 

  
**Checklist**

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] Any dependent changes have been merged and published

NeilBarton-NOAA committed:
3DVAR IAU options

NeilBarton-NOAA committed:
Merge branch 'NOAA-EMC:develop' into ATM_3DVAR_IAUT

NeilBarton-NOAA committed:
Merge branch 'NOAA-EMC:develop' into ATM_3DVAR_IAUT

NeilBarton-NOAA committed:
Merge branch 'NOAA-EMC:develop' into ATM_3DVAR_IAUT

aerorahul commented on parm/config/config.base.emc.dyn:
```diff
@@ -346,6 +332,28 @@ if [ $DOHYBVAR = "YES" ]; then
         export FHOUT_ENKF=3
     fi
+else
+  export DOIAU=${DOIAU:-"NO"}   # 3DVAR defaults to no IAU
```
If it is being set as default, then there is no reason to have `${DOIAU:-NO}`
```suggestion
  export DOIAU="NO"   # 3DVAR defaults to no IAU
```

aerorahul commented on parm/config/config.base.emc.dyn:
```diff
@@ -346,6 +332,28 @@ if [ $DOHYBVAR = "YES" ]; then
+fi
+
+# IAU related parameters
+if [[ $DOHYBVAR == "NO" && $DOIAU == "YES"  ]]; then
+  export IAUFHRS="6"
+  export IAU_FHROT="3"
+  export IAU_FILTER_INCREMENTS=.true.
+else
+  export IAUFHRS="3,6,9" # Enable 4DIAU or control with 3 increments
+  export IAU_FHROT=$(echo $IAUFHRS | cut -c1)
+fi
```
This if-block is a bit confusing, as both, the if and else are for `DOIAU=YES`.
The if-block is for 3DVar with IAU and the second block can be hybrid (3D or 4D), but is only going to work for 4DHybEnVar as 3DEnVar will not produce increments at hours 3 and 9.

NeilBarton-NOAA replied:
yes, the changes in the if block are for 3DVAR with IAU. 3DVAR with IAU does not cycle without these changes.  @CatherineThomas-NOAA could you comment on the else part of the block as I'm not familiar with the 3DEnVar settings?  

A proposed change is 

if [[ $DOHYBVAR == "NO" && $DOIAU == "YES"  ]]; then
  export IAUFHRS="6"
  export IAU_FHROT="3"
  export IAU_FILTER_INCREMENTS=.true.
fi
export IAUFHRS=${IAUFHRS:-"3,6,9"}
export IAU_FHROT=${IAU_FHROT:-$(echo $IAUFHRS | cut -c1)}

CatherineThomas-NOAA replied:
The IAU settings for 3DEnVar will be similar to 3DVar, just set the *_ENKF IAU variables the same as their deterministic counterparts (IAUFHRS --> IAUFHRS_ENKF, etc).

aerorahul commented on parm/config/config.base.emc.dyn:
```diff
@@ -346,6 +332,28 @@ if [ $DOHYBVAR = "YES" ]; then
+fi
+export IAU_DELTHRS=6
+export IAU_OFFSET=6
+export DOIAU_ENKF=${DOIAU:-"YES"}   # Enable 4DIAU for EnKF ensemble
+export IAUFHRS_ENKF="3,6,9"
+export IAU_DELTHRS_ENKF=6
```
These should be wrapped in an if-block such as:
```
if [[ $DOHYBVAR = "YES" ]]; then
    export DOIAU_ENKF=${DOIAU:-"YES"}   # Enable 4DIAU for EnKF ensemble.
    export IAUFHRS_ENKF="3,6,9"
    export IAU_DELTHRS_ENKF=6
fi
```

NeilBarton-NOAA replied:
updated

CatherineThomas-NOAA replied:
This IF block is not sufficient.  The IAU*ENKF variables should only be set to these values if DOIAU_ENKF="YES" and for 4D.

aerorahul commented on parm/config/config.base.emc.dyn:
```diff
@@ -338,6 +323,7 @@ export lwrite4danl=".true."
 
 # EnKF output frequency
 if [ $DOHYBVAR = "YES" ]; then
+    export DOIAU="YES"        # Enable 4DIAU for control with 3 increments
```
This can be for 3DIAU or 4DIAU.  `DOHYBVAR` is a key for enabling hybrid regardless of 3D or 4D.
The choice of 3D or 4D is deferred to `l4densvar`.

Is the goal here to turn on IAU for hybrid (3D or 4D) as the default?.  If so, then
```suggestion
    export DOIAU="YES"
```

aerorahul reviewed (commented):
Most of this looks good to me. I just have a few comments inline.
I would like to ask someone from the DA team to review these changes so they are not breaking the core functionality.
Do you have logs of the tests you ran for:
3DVar w/ IAU
3DVar w/o IAU
4DEnVar w/ IAU (this is the default)

NeilBarton-NOAA committed:
Update parm/config/config.base.emc.dyn

correct

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

NeilBarton-NOAA committed:
Update parm/config/config.base.emc.dyn

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

NeilBarton-NOAA committed:
add if statement for clarity and add {} for syntax

WalterKolczynski-NOAA commented on parm/config/config.base.emc.dyn:
```diff
@@ -346,6 +332,30 @@ if [ $DOHYBVAR = "YES" ]; then
+fi
+
+# IAU related parameters
+if [[ ${DOHYBVAR} == "NO" && ${DOIAU} == "YES"  ]]; then
```
As constructed, this will never be true. L336 is setting `$DOIAU` to NO if `$DOHYBVAR` is not YES.

NeilBarton-NOAA replied:
The user may change the settings after setup_expt.py and, if so, the IAFHRS and IAU_FHROT need to be set as in lines 341 and 342 to cycle.  In develop, DOHYBVAR=NO defaults to IAU=YES, which does not cycle out-of-the-box, and, to my understanding, is not as well tested as DOHYBVAR=NO and IAU=NO.

NeilBarton-NOAA committed:
simplifying the changes for IAU and 3DVAR*
" Check if cycle is cold starting, DOIAU off, or free-forecast mode" if statement moved after Hybrid variables as it's needed for new if statement

NeilBarton-NOAA commented:
I just submitted changes to config.base.emc.dyn that simplifies the changes. The change is the addition of below if block after the "#Hybrid related" line as DOHYBVAR needed to be defined. These changes set the variables if the user runs 3DVAR with IAU. 

# if 3DVAR and IAU
if [[ ${DOHYBVAR} == "NO" && ${DOIAU} == "YES"  ]]; then
     export IAUFHRS="6"
     export IAU_FHROT="3"
     export IAU_FILTER_INCREMENTS=.true.
     export IAUFHRS_ENKF="6"
fi

In addition, the "if [[ "$MODE" = "cycled" && "$SDATE" = "$CDATE" && $EXP_WARM_START = ".false." ]] ||"
had to be moved to after the if statement so IAU_OFFSET and IAU_FHROT could be reset when needed.

CatherineThomas-NOAA commented on parm/config/config.base.emc.dyn:
```diff
@@ -332,6 +327,7 @@ export DO_JEDIENS="NO"
 # Hybrid related
 export DOHYBVAR="YES"
 export NMEM_ENKF=@NMEM_ENKF@
+export NMEM_EFCS=30
```
@NeilBarton-NOAA I think this line and Line 339 were accidentally added in your commit.

NeilBarton-NOAA replied:
This is from the current develop branch. I used the latest develop file for changes and the last commit. The same with 

export FHMAX_ENKF_GFS=120

Should I remove?

aerorahul replied:
This was added recently to allow a subset of members to be run as part of `gfsefmn` for GEFS.

CatherineThomas-NOAA commented on ush/parsing_namelists_FV3.sh:
```diff
@@ -396,6 +396,7 @@ if [ $DOIAU = "YES" ]; then
   iau_delthrs  = ${IAU_DELTHRS}
   iau_inc_files= ${IAU_INC_FILES}
   iau_drymassfixer = .false.
+  iau_filter_increments = ${IAU_FILTER_INCREMENTS:-.false.}
```
.false. should probably be ".false." for consistency

NeilBarton-NOAA replied:
added

NeilBarton-NOAA committed:
adding quotes around .true. and .false.

CatherineThomas-NOAA commented:
@NeilBarton-NOAA Thanks for these changes.  I think this is much cleaner and covers the main hybrid 4DEnVar and 3DVar cases.

WalterKolczynski-NOAA commented:
Waiting for @CatherineThomas-NOAA to formally give an approving code review.
